nexus/blueprint-execution/src/dns.rs

```diff
@@ -0,0 +1,744 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Propagates internal DNS changes in a given blueprint
+
+use crate::Sled;
+use dns_service_client::DnsDiff;
+use internal_dns::DnsConfigBuilder;
+use internal_dns::ServiceName;
+use nexus_db_model::DnsGroup;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::datastore::DnsVersionUpdateBuilder;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::Blueprint;
+use nexus_types::deployment::OmicronZoneType;
+use nexus_types::internal_api::params::DnsConfigParams;
+use omicron_common::address::get_switch_zone_address;
+use omicron_common::address::CLICKHOUSE_KEEPER_PORT;
+use omicron_common::address::CLICKHOUSE_PORT;
+use omicron_common::address::COCKROACH_PORT;
+use omicron_common::address::CRUCIBLE_PANTRY_PORT;
+use omicron_common::address::CRUCIBLE_PORT;
+use omicron_common::address::DENDRITE_PORT;
+use omicron_common::address::DNS_HTTP_PORT;
+use omicron_common::address::MGD_PORT;
+use omicron_common::address::MGS_PORT;
+use omicron_common::address::NEXUS_INTERNAL_PORT;
+use omicron_common::address::NTP_PORT;
+use omicron_common::address::OXIMETER_PORT;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::Generation;
+use omicron_common::api::external::InternalContext;
+use slog::{debug, info, o};
+use std::collections::BTreeMap;
+use uuid::Uuid;
+
+pub async fn deploy_dns<S>(
```

Comment:
Two tiny nits:

* I'd make this `pub(crate)`
* Given the former, I think you could drop `<S>` and take `creator: String`, and let the caller do the conversion. That keeps the ergonomics of the public API without needing the generic on this inner function.